core/shared/src/main/scala/zio/logging/internal/json/JsonValidator.scala

```diff
@@ -0,0 +1,165 @@
+package zio.logging.internal.json
+
+import scala.annotation.tailrec
+
+object JsonValidator {
+
+  def stringIsJson(string: String): Boolean = {
+    val startIdx = stringCouldBeJson(string)
+    if (startIdx == NOT_A_JSON) false
+    else checkJson(string, startIdx)
+  }
+
+  private def checkJson(
+    string: String,
+    startIdx: Int
+  ): Boolean = {
+    var stack        = SmallBooleanStack()
```

Comment:
The trick to eliminate use of `SmallBooleanStack` is to _recurse_: this allows you to store on the JVM stack the information on the block type. "Popping" the stack corresponds to exiting one of the recursions, and bubbling up to the next higher-level; "pushing" a boolean on the stack corresponds to recursing.

In order to implement this optimization, you can generalize `checkJson` to take the current state and the boolean. 

Note that when completed, this means we cannot validate JSON of arbitrary depth. We could solve that by switching over to a slower-based validator after we recurse `n` times (where `n` > 200, say). But I'd say solving that is not essential for v1.

Reply:
Done.